PartDesign: thread texture

alfrix opened the pull request:
I think I have everything working, the thread pattern was drawn by me on inkscape, it supports tapered threads, even thread depth, it adjusts the pattern depending on the pitch, the texture is translucent so the material color can be shown through, I've had to enable texture in the view provider extension, and adjust the triangulation function because it was giving me issues with bad triangles.

I'm working on this because of the bounty, I need the money to get a new laptop so I can compile with a more reasonable turnaround, it takes me a long time to rebase sometimes hours because of my i5 430m (yes that old), I don't know if this is good enough, but at least I'm trying.

closes #5567

Screenshots:
| depth of 35mm | |
| --- |--- |
| <img width="482" height="517" alt="image" src="https://github.com/user-attachments/assets/0fbf4770-c5c9-4123-acb8-d23061555c3a" /> | <img width="265" height="219" alt="image" src="https://github.com/user-attachments/assets/2bf5cca3-2410-4a9a-bdd5-c30d042c7065" /> |

disable feature
<img width="265" height="219" alt="image" src="https://github.com/user-attachments/assets/705deb0c-57a7-4571-81af-15ff74ecfc89" />

hyarion commented:
Is read depth correctly implemented? The screw looks difficult to fasten if the treads don't start from the tip:

![image](https://github.com/user-attachments/assets/7e148679-37c3-4949-8389-6d6ece2f8b28)

Looks cool though 👍

maxwxyz commented:
@FreeCAD/design-working-group @FreeCAD/cad-working-group FYI

pierreporte commented:
This is huge! I'm in holidays right now and for then next two weeks so I may not be able to evaluate this PR properly until then.

@alfrix do you plan to do the whole thread feature overhaul after this PR?

hyarion commented:
A big part of the issue seems to be techdraw support, is this something you're planning to do in this PR too?

hyarion commented:
> > A big part of the issue seems to be techdraw support, is this something you're planning to do in this PR too?
> the issue is about adding a thread texture into PartDesign, so no this is not part of this PR, i've never looked at TD code at the moment, so idk how hard would that be, but it would be a different PR for sure

Im referring to this comment which further describes the desired feature: https://github.com/FreeCAD/FreeCAD/issues/5567#issuecomment-2543142577

alfrix commented:
> > > A big part of the issue seems to be techdraw support, is this something you're planning to do in this PR too?
> 
> > the issue is about adding a thread texture into PartDesign, so no this is not part of this PR, i've never looked at TD code at the moment, so idk how hard would that be, but it would be a different PR for sure
> 
> Im referring to this comment which further describes the desired feature: [#5567 (comment)](https://github.com/FreeCAD/FreeCAD/issues/5567#issuecomment-2543142577)

That comment describes everything that is missing/could be implemented, funnily enough i've already implemented some of that already (new UI, added standards BSW BSF, NPT, NPS, fixed modeled tapered, etc)


> I presume you've already explored a texture that is angled (something like the attached) that when wrapped that gives the illusion of real threads? Was there some kind of issue?
> 
> ![ThreadOverlayPitched](https://private-user-images.githubusercontent.com/17993710/468328572-df9b6572-16dc-450d-b432-eb9f243a48f1.svg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTI5NzY5NDIsIm5iZiI6MTc1Mjk3NjY0MiwicGF0aCI6Ii8xNzk5MzcxMC80NjgzMjg1NzItZGY5YjY1NzItMTZkYy00NTBkLWI0MzItZWI5ZjI0M2E0OGYxLnN2Zz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzIwVDAxNTcyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBhOThkYjY3M2M3MDY0N2U0NWNlMDQ1YTgxNWUwNzZjMjExYjBjNTBiNmRkMjFhMzBjZWI3ZDFiYjBjYTczYzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.VXCf3u2EcamOmiM_8dyY7BTsztuiG3UeIkCmSkk5-1M)

i have, the problem is that you would create a seam that will look bad
<img width="500" height="380" alt="image" src="https://github.com/user-attachments/assets/d521fd25-1513-4284-8da5-c66919e436c9" />

belinski-roman commented:
In Curves WB, you can create a sketch on a cylindrical surface. Essentially, the imitation of a thread is a diagonal line on a rectangle (the unfolding of a cylinder). By looking at the code in that workbench, you may be able to solve the problem of the strange seam

alfrix commented:
i managed to fix the seam with [6dfe56a](https://github.com/FreeCAD/FreeCAD/pull/22573/commits/6dfe56ac3cba3628ce90c7f73646ce5f1e9b2c3d) and made possible the diagonal texture

Transparent view:
<img width="312" height="353" alt="image" src="https://github.com/user-attachments/assets/0598ce25-7669-4c46-9edd-78eabaf9dde7" />

and the thread direction is now shown too.

|right hand | left hand |
|-|-|
| <img width="312" height="353" alt="image" src="https://github.com/user-attachments/assets/1ebcf9b5-c2f9-4eab-b9b6-b9de7f38fa65" /> | <img width="312" height="353" alt="image" src="https://github.com/user-attachments/assets/94c5baa8-fde3-4ccf-84e9-5f551fc8a33c" /> |

pierreporte commented:
I just tried the PR and I must say it’s impressive. It understands pitch and direction, and color changes automatically. Good job @alfrix!

Now, a few comments:

1\) I’m not sure if the cosmetic thread should correspond to hole type. In my opinion, that list should still have the same choices, but presented a bit differently: threaded, pre-drilled, and clearance. When threaded is selected, there is a checkbox appearing to make the thread modeled (unchecked by default, but beware of backwards compatibility). When pre-drilled is selected, the cosmetic thread is not added and it behaves like in current FreeCAD. No change for clearance holes.

2\) There are now 2 parameters in the hole feature property view: `Model Thread` and `Cosmetic Thread`. They should be merged into one: the thread is either cosmetic or modeled. Currently it can be both or neither while `Threaded` is set to `true`.

3\) There should be a set of options in the settings window and display settings in the toolbars, so that cosmetic threads visibility can be adjusted. In the settings, users chooses if they want them displayed all the time (default), never, or hidden when the total number in the view exceeds a certain amount (user-set value with a sane default like 200 perhaps). On the toolbar, the same can be done but just for the current view if users want to temporarily change from default behavior. @FreeCAD/design-working-group could weigh in.

pierreporte commented:
Ping @FreeCAD/cad-working-group

alfrix commented:
> 1) I’m not sure if the cosmetic thread should correspond to hole type. In my opinion, that list should still have the same choices, but presented a bit differently: threaded, pre-drilled, and clearance. When threaded is selected, there is a checkbox appearing to make the thread modeled (unchecked by default, but beware of backwards compatibility). When pre-drilled is selected, the cosmetic thread is not added and it behaves like in current FreeCAD. No change for clearance holes.
> 2) There are now 2 parameters in the hole feature property view: `Model Thread` and `Cosmetic Thread`. They should be merged into one: the thread is either cosmetic or modeled. Currently it can be both or neither while `Threaded` is set to `true`.
> 3) There should be a set of options in the settings window and display settings in the toolbars, so that cosmetic threads visibility can be adjusted. In the settings, users chooses if they want them displayed all the time (default), never, or hidden when the total number in the view exceeds a certain amount (user-set value with a sane default like 200 perhaps). On the toolbar, the same can be done but just for the current view if users want to temporarily change from default behavior. 

IMHO the geometry created takes precedence, and the current combo choices are self-explanatory for what geometry you get without adding extra clicks, i would prefer to add "Tap Drill (Textured)" or "Tap Drill / Cosmetic Thread" that would alleviate the need of an extra preference option

> I just found a nasty bug. When removing material from the hole with a different feature afterwards, the texture remains in the 3D space. It should only be kept where the thread remains.

i'll investigate that

pierreporte commented:
@alfrix

> IMHO the geometry created takes precedence, and the current combo choices are self-explanatory for what geometry you get without adding extra clicks, i would prefer to add "Tap Drill (Textured)" or "Tap Drill / Cosmetic Thread" that would alleviate the need of an extra preference option

Actually, the cosmetic threads should be treated as if the thread was really there. From the user’s point of view, the part is threaded. The representation is only simplified. Tap-drill (or pre-drill) is what is is: just a hole. So really the choice is between threaded and not threaded, and when it is threaded the representation can be complete (modeled) or simplified (cosmetic). It’s the same in drawings: threads are almost never represented as they are physically because it doesn’t make sense, yet the part is definitely threaded. In both case, the thread is not different: it’s exactly the same physically at the end but its representation within the computer (or on parper) can differ.

Maybe part of the problem is that there is the tap-drill option in the first place: it’s not really useful. Only time you do it, usually you have a specific diameter in mind which is not the one defined by the feature. But there’s no need to strip it completely from FreeCAD.

> > Im referring to this comment which further describes the desired feature: https://github.com/FreeCAD/FreeCAD/issues/5567#issuecomment-2543142577
>
> That comment describes everything that is missing/could be implemented, funnily enough i've already implemented some of that already (new UI, added standards BSW BSF, NPT, NPS, fixed modeled tapered, etc)

What you did was really nice but it was more of an adaptation of the old feature while retaining most of its quirks. The UI proposal is about much deeper changes on how the threads are selected. It wasn’t really planned to have this intermediate state but we worked together to make it as close as possible to the initial intent.

It would be really nice if you wanted to go further, but it can be done in new pull requests.

------------------

In https://github.com/FreeCAD/FreeCAD/issues/5567#issuecomment-2576181931 there is an illustration of dashed circles that help visualizing the actual extent of the thread. Do you think it’s doable in this PR? Note that it should adapt to the surface, so in the picture I posted yesterday showing a bug, it should be moved on the conical surface, but in the general case it could be any kind of surface. In the case of threaded holes that go through the material, and where the whole length is threaded, the same circle should be put on the exit surface with the same requirements.

pierreporte commented:
Another variation of the bug. I’ve removed material on the block to make a step that partially cut the thread. The thread should be only on the remaining material.

Note that the proposed dashed outline should extend on the vertical face, but it would be two lines.

<img width="1274" height="670" alt="image" src="https://github.com/user-attachments/assets/aef53639-3a9a-4640-969e-cf59ac4fe6a8" />

hyarion commented:
I don't see any reason to use XPM,  we're moving away from it as they are hard to work with and only really works for low res icons, which is a problem on high-dpi screens.

If anything it would be better to use the SVG file directly (load svg to an image which then can be used as a texture). I think the current solution to have pre-rendered versions is acceptable too.

pierreporte commented:
Ping @kadet1090 for the potential problem with previews.

I'll create an issue to request finite length modeled threads. It will be P1.

hyarion commented:
Added it to 1.2 as 1.1 is so close and there's a high risk that this will add some regressions.
It is an awesome feature and I'm looking forward using it when its done :)

alfrix commented:
Fixed in the last commit
* body disappears on first feature added afterwards
* texture gets angled if multiple holes on same feature
* show thread-depth in the taskpanel
* react on the fly to  depth or direction changes
* texture is not cleared when body/features visibility changes
* the hole needs to be updated with the last visible feautre shape not with the last body

alfrix commented:
with the last 2 commits i've fixed
* threaded face cannot be selected
Known issues
* if the body is translucent texture changes opacity at certain angles
* thread depth plane not cutting (maybe needs testing)
i'm not fully sure but i think it might be a driver issue where the default for the depth buffer cause issues with the clipping plane, i've pushed a commit that forces the depth to 24 bit, please test and let me know the gpu used

i didn't fully understand the thread length issue, the values are not reset? if that's the case then the bug is on main and not particular to this PR

pierreporte commented:
> i didn't fully understand the thread length issue, the values are not reset? if that's the case then the bug is on main and not particular to this PR

When hole depth is through all, then yes it is reset. It is not reset when the hole depth is finite. Modeled thread work as expected in all cases.

alfrix commented on src/3rdParty/OndselSolver:
thanks, this submodules sometime slip in after a rebase if i ammend, i'll remove it

pierreporte commented:
It works!

<img width="975" height="718" alt="image" src="https://github.com/user-attachments/assets/945e90d4-c0d4-4495-bf7b-c628eea9163a" />

There is still the issue when hole depth is through all, that resets the thread depth to hole depth after task validation. It is needed to use the property view to adjust manually the value (the above picture was made in this situation). When using modeled threads there is no reset.

pierreporte commented:
Thanks for the new PR. I have still a thing or two to test which I forgot last night. Apart of potential new issues after the remaining tests, there is no line indicating the end of the thread. Could you add one?

Now the UI is the biggest thing.

The length management of modeled threads is problematic IMO. It tries to mimic the runout but it doesn't give it the right geometry (it's a separate issue). It is also not shared with cosmetic threads despite being a legitimate value (in Catia it can be set). Ideally their should be an optional field for both kind of threads, and the extra length added to modeled threads should be this value. I don't know if there are standard values for the default. Ping @FreeCAD/cad-working-group 

The dashed outline I mentioned in earlier comments would make use of the value so it's important that the parameter is shared with both modes.

Have you found a way to make the dashed outline? It doesn't need to be part of this PR. If it is merged without it I'll create an issue.

pierreporte commented:
I’ve tested making two coaxial holes on a cylinder, each from one of the end, to see if the thread is still there. I didn’t test with through all holes since #23695 is not merged yet.

Cylinder length : 20 mm, thread depth 6 mm.

Hole depth 12 mm on each side, so that the cylinder is drilled through, but in two steps: the two thread textures disappear even though the second one never touches the first thread (but it does touch the first hole). File: [coaxial holes.zip](https://github.com/user-attachments/files/22186277/coaxial.holes.zip) (rename to .fcstd). Funny enough, when I cut the part with a subtractive cut to make a picture to post here, the second thread appears (but not the first). File: [coaxial holes cut.zip](https://github.com/user-attachments/files/22186283/coaxial.holes.cut.zip) (rename to .fcstd).


<img width="902" height="486" alt="image" src="https://github.com/user-attachments/assets/1fd020d8-653d-40f0-b5c1-af99f82acbec" />

Hole depth 8 mm on each side, so there is no intersection at all. It is where it gets funny: without the cut only the first thread appear, and with the cut only the second one appear. Files: [coaxial holes 2.zip](https://github.com/user-attachments/files/22186302/coaxial.holes.2.zip), [coaxial holes cut 2.zip](https://github.com/user-attachments/files/22186306/coaxial.holes.cut.2.zip) (rename to .fcstd).

<img width="903" height="486" alt="image" src="https://github.com/user-attachments/assets/da49247f-b39c-481d-b917-d2c1ab613184" />

<img width="886" height="526" alt="image" src="https://github.com/user-attachments/assets/2a6d0778-8095-4244-b2e8-fd9a90bd6db8" />

When the threads are not created fully inside the part, the same problem appears, but the threads switching visibility happens only when the cutting operation touch the threads.

<img width="1315" height="870" alt="image" src="https://github.com/user-attachments/assets/5f8a873f-4282-4122-8b7e-6431d0bfb6f9" />

----------------------

In terms of usability, I would expect that drilling along the axis of an existing thread, with the same or smaller diameter that the pre-hole (so basically if I set the hole depth to 20 mm or more) shouldn’t make the thread disappear. In real life it may not work but it makes design easier by allowing not to take care of those details (only the machinist will have to).

If the two threads were to have an intersection, then I think that displaying the intersection of texture would be an indicator that there is a problem. Same in TechDraw (for the future).

alfrix committed:
PD: Hole Thread texture

alfrix commented:
The other coaxial issue should be fixed, the issue was that if 2 holes intersect each other completely, then the surface type chagens to a trimmed one that needs to be unwrapped

Added an offset/phase so intersecting threads can be clearly seen 

refactored into smaller functions and cleanup includes

alfrix commented:
> * a solid line (like an edge) at the end of the thread to show precisely where it is

i think i can manage that in a day or 2

> * a dashed outline around the thread to show its true extent.

at the moment it doesn't occur to me how to achieve that.

> Now the biggest remaining issue is the UI. As it is now, the option shouldn’t be called “Cosmetic thread / Tap drill” but just “cosmetic thread”. But ideally, the whole selector should be changed, so that the hole type could be the 3 following options only: “Clearance / Passthrough”, “Threaded”, “Tap-drill”. Then, when “Threaded” is selected, there is a new selector to give the option to have a modeled thread or a cosmetic thread (cosmetic should be the default).

i'll rename it to remove Tap Drill, i guess the user can notice it wen changing modes, but i want to leave both cosmetic and modeled as is in the combo-box i think it's more direct and clear rather than creating yet another thing.

> This is linked to an issue with the property view, because when you make a modeled thread, and toggle the “Cosmetic Thread” property on, then the modeled thread disappears. I still think that the two properties “Modeled Thread” and “Cosmetic Thread” should be merged into one, and deactivated when the “Threaded” property is off.

it's not really related to this PR, but i do agree that we can improve the property list situation. the modeled thread has a recomputation blocker  unless you tick "update thread view" and that can be weird in most situations, the ideal solution would be to remove that blocker, after either improving the modeled performance to a bearable threshold and  or at least have a "recomputing" bar that can be canceled.

pierreporte commented:
@alfrix please note that the hole task panel has recently changed. you should build your PR upon this new UI:

<img width="432" height="820" alt="image" src="https://github.com/user-attachments/assets/57b74e75-91c1-4442-b682-c5fd1fb8b919" />

Also, you may need a migration script if you change the structure of properties, so that old files can still be opened.

pierreporte commented:
My reasoning was more about function than manufacturing processes. But even then, it is still two kinds of threads, not two kinds of holes. Also, listing thread types as holes types blocks you from specifying a default thread type, because the two are necessarily equal in your UI. It means that users are going to thing that both are valid most of the time, while good practice dictates that cosmetic threads should always be used, unless the manufacturing process requires the threads to be physically in the model.

pierreporte commented:
I forgot the issue mentioned in https://github.com/FreeCAD/FreeCAD/pull/22573#issuecomment-3255336529, that the texture keeps its color when the thread is selected. Is it possible to change it dynamically?

alfrix commented:
added some std::optionals for robustness
rebased against current main
the color of both the texture and marker is 50% transparent black, it will let body color see-through
the deflection of the marker is an issue, i cannot replicate the resolution of the body exactly, that is why it's offset inwards, it can be closer, i'll place it as close as possible for the default deviation

alfrix commented:
> Is it possible to draw the circle using a texture...

mmm, maybe with a secondary clip plane,  i'll try that out

alfrix commented:
> The new texture appears to be missing from source. Did you forget to add it to git?

hahaha yes i did good catch, i'll make it thinner and darker

pierreporte commented:
The line is perfectly crisp, but still constant size and fixed color.

<img width="1274" height="693" alt="image" src="https://github.com/user-attachments/assets/c3d33281-8976-4807-a373-4f4d6b2f82b2" />

alfrix commented:
> The line is perfectly crisp, but still constant size and fixed color.

those are limitations of using a texture, we either make it semi-transparent or it's a solid color, you didn't want the semi-transparent nor the solid one, also the size behaves differently to lines in the model, that's why i used a soline in my first attempt as it would behaves more to what you expect...

for now i'm fixing an ui issue i detected, where the thread clearance is not properly shown and other small issues

pierreporte commented:
Considering the actual 3D circle again, is it possible to determine a mathematical relation between the hole apparent diameter and the circle apparent diameter, and use it to apply a scale factor to the circle?

alfrix commented:
> Considering the actual 3D circle again, is it possible to determine a mathematical relation between the hole apparent diameter and the circle apparent diameter, and use it to apply a scale factor to the circle?

the size of the circle is not really the problem, the actual issue is the deviation value, we need to match the polygonal shape of the hole to make it as close as possible, but that is something i could not achieve, i've tried calculating the sagita derived from the angular deflection of the settings, but that doesn't seem to be respected, the hole shape is only affected by the deviation value, which is supposedly a percentage of the bounding box, but no matter what i tried i cannot calculate the appropriate value, i even made a spreadsheet of the values that works, but i cannot reach a formula to match them.

without the proper deflection value if the circle is too close it will get in and out of the body due to the polygons having different number of sides

0,5% is the default deviation in freecad, here is the map i made there ranges that work
<img width="861" height="296" alt="image" src="https://github.com/user-attachments/assets/3004750f-8d86-4de1-a5ca-cdbf1b17d0eb" />

alfrix commented:
> Then I propose to put this part of the PR aside for a follow-up. It will likely be a problem with the dashed outline as well. So what about having a complete review of just the texture and UI, so it can be merged right after the branching if we’re lucky?

ok, removed end of thread marker until i can match the deflection

alfrix commented:
> Is the PR ready for review now?

i think so, the texture part which is the main feature is ready

pierreporte commented:
I had an idea after thinking about a design we made a while back at my job. Basically there is a M12 threaded hole that is needed for a plug. Inside this hole, we had to fit with tight fitting tolerances, that was a tiny bit larger than the tap drill diameter of the thread. We adjusted the tap drill diameter directly with the hole feature, but today this struck me: if you drill through the thread, but with a diameter smaller than the outer thread diameter, the thread will still exist even though the profile is incomplete. In the drawing, the hole should appear threaded. It is why I think that the texture should stay if the hole is increased by further operations, unless the resulting diameter is larger that the outer diameter, that would correspond to the dashed line.

@alfrix Do you want it to be a follow-up?

alfrix commented:
> Is there an issue corresponding to your forum post?
> 
> What I wrote about is a different thing: you have an existing hole (that may or may not have the standard tap drill diameter), and then drill again at a different diameter, to shave the thread, but not removing it completely.
> 
> It can be a follow-up PR no problem, I’ll create the corresponding issue once this one is merged.
> 

i don't seem to have created a formal issue in the repo, i wanted so forum feedback, but as you see there wasn't any.

regardless to what you would do in real-life to achieve it, in the model it really doesn't make sense to do that, it would make more sense to just do the model with the final tap-drill diameter, so the bore recognition code doesn't have to more complex than it is, even if at some point in the future we tag the face using a robust TopoNaming instead of the current find logic, that face will disappear with a second operation, leaving you again with the same issue of trying to track a face that doesn't exist.
> 
> i say it doesn't make sense to try catch a second hole operation, but it does make sense to adjust the tap-drill to it's final diameter

maxwxyz commented:
@alfrix ping for the conflicts

Reqrefusion commented:
@alfrix, if you send your account information to FPA when you return, we can quickly send you the relevant bounty. We apologize for keeping you waiting so long.

kadet1090 commented on src/Mod/PartDesign/Gui/ViewProviderBody.h:
```diff
@@ -108,8 +111,18 @@ class PartDesignGuiExport ViewProviderBody: public PartGui::ViewProviderPart,
+    std::map<const PartDesign::Hole*, SoSwitch*> m_threadOverlays;
+    void clearThreadTextures();
+    void afterRecompute(const App::Document&, const std::vector<App::DocumentObject*>& recomputedObjs);
+    boost::signals2::scoped_connection m_RecomputedConn;
```
This pretty much makes body aware about a specific feature and moves a bit of logic to it. The body should not know about implementation details of it's children as this violates open-closed principle and single responsibility principle. This is a rather serious architecture issue which should be addressed.

Ideally information like that should simply be stored within geometry and then simply handled on render level. That however probably is not possible so we need to figure something out here - I wonder if we can create some more generic mechanism for that? For example an extension to a view provider of given feature that could add additional "persistent" coin nodes? This way it would no longer be about thread textures specifically but any other feature could add annotations like that in the same manner. 

This would be much cleaner from architecture standpoint as the body will still only know that it has a certain set of features, that can simply affect body in one more way.

@FreeCAD/code-quality-working-group - thoughts?

kadet1090 commented on src/Mod/PartDesign/Gui/ViewProviderHole.cpp:
```diff
@@ -68,3 +108,553 @@ TaskDlgFeatureParameters* ViewProviderHole::getEditDialog()
+        static_cast<float>(holeNormalAxis.X()),
+        static_cast<float>(holeNormalAxis.Y()),
+        static_cast<float>(holeNormalAxis.Z())
+    );
```
```suggestion
    SbVec3f endPlanePoint = Base::convertTo<SbVec3f>(endPlanePnt);
    SbVec3f endPlaneNormal = Base::convertTo<SbVec3f>(holeNormalAxis);
```

    SbVec3f endPlaneNormal(
        static_cast<float>(holeNormalAxis.X()),
        static_cast<float>(holeNormalAxis.Y()),
        static_cast<float>(holeNormalAxis.Z())
    );

```

kadet1090 commented on src/Mod/PartDesign/Gui/ViewProviderHole.cpp:
```diff
@@ -68,3 +108,553 @@ TaskDlgFeatureParameters* ViewProviderHole::getEditDialog()
+        return std::nullopt;
+    }
+
+    return gp_Dir(normal.x, normal.y, normal.z);
```
```suggestion
    return Base::convertTo<gp_Dir>(normal);
```

kadet1090 commented on src/Mod/PartDesign/Gui/ViewProviderHole.cpp:
```diff
@@ -68,3 +108,553 @@ TaskDlgFeatureParameters* ViewProviderHole::getEditDialog()
+            return posteriorFeature->Shape.getValue();
+        }
+    }
+    return body->Shape.getValue();
```
I don't really understand what is purpose of this code and it is repeated in the body - can you elaborate on it?

Reqrefusion commented:
@alfrix In order for us to send you the bounty here, you need to send your account information to accounting@freecad.org. Hopefully, it will be resolved as soon as possible.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

alfrix committed:
refactor: vp-texture remove logging

alfrix committed:
refactor: vp-hole cleanup unused declaration

alfrix committed:
refactor: vp-hole remove logging

alfrix committed:
refactor: vp-hole use convertTo

alfrix committed:
refactor: vp-hole use freecad_cast

alfrix committed:
refactor: make unwrapsurface private

alfrix committed:
fix: logic error with model-thread handling

alfrix committed:
fix: modelthreadchanged issues

threaddepth doesnt depend on modeled or not so remove it

alfrix committed:
fix: model thread getting disabled when it shouldnt

alfrix committed:
revert: use base-0 poly triangulation

previously had issues, but all is good now

alfrix commented:
> Ah it ate up my comment...
> 
> While testing if it works after solving conflicts I found one issue:
> 
> If you create a hole that goes through the whole shape and then extend it - the thread will be extended too, which is not expected. To reproduce:
> 
>     1. Create some flat shape, rectangle perhaps and extend it upwards 10mm
> 
>     2. Create a hole on the top face, set the depth to 20mm
> 
>     3. Pad the bottom face downwards 20mm.
> 
> 
> The thread will extend 20mm instead of expected 10. It probably should capture the geometry after creation and use it on later stages so it does not get modified.
> 
> While trying to find if that is maybe a small issue in code I found few more issues, including one relatively big one that technically changes contract of one of the functions which I have not noticed due to how diffs works. The get triangulation function outputs the `n1` etc to output parameter so it technically changes contract. Not sure if it will for sure break something, but it has the potential to do so.
> 
> Given that I re-reviewed it and found few more relatively minor issues that might be worth to fix before merging.

I understand, but this is actually the intended behavior, because threadDepth was left at 20mm, the system is simply honoring that value once the geometry exists to support it.
The texture mesh it's evaluated based on the final resolved geometry, not intermediate history steps. Trying to 'clip' a texture based on a previous state in the tree would make the feature unpredictable, would need to save meshes for every step and intersect them at the end to apply the texture.

If the goal is a 10mm thread, the threadDepth parameter just needs to be set to 10mm, i genuinely think that most users would be confused and would not understand why their thread is behaving weirdly because intermediate geometry, not counting the additional state related bugs, complexity and performance impact that implementing that would have.

alfrix commented on src/Mod/Part/App/Tools.cpp:
```diff
@@ -310,9 +310,12 @@ bool Part::Tools::getTriangulation(
-        --n3;
+
+        // OpenCascade's Poly_Triangulation::Triangle().Get()
+        // returns 1-based indices, and Poly_Triangle (used in 'facets') is expected to store
```
[ec1560d](https://github.com/FreeCAD/FreeCAD/pull/22573/commits/ec1560de83181c0b197b3c31ac51dc27062f8c8f)

alfrix commented on src/Mod/PartDesign/Gui/ViewProviderHole.cpp:
```diff
@@ -70,3 +135,602 @@ TaskDlgFeatureParameters* ViewProviderHole::getEditDialog()
+    std::string theadDepthType = pcHole->ThreadDepthType.getValueAsString();
+    if (theadDepthType == "Hole depth") {
+        m_endThreadClipper->on = FALSE;
+        Base::Console().message("Thread clip plane disabled\n");
```
[5bde484](https://github.com/FreeCAD/FreeCAD/pull/22573/commits/5bde484a056dd75beab72153888755b8a8f93072)

alfrix commented on src/Mod/PartDesign/Gui/ViewProviderHole.cpp:
```diff
@@ -70,3 +135,602 @@ TaskDlgFeatureParameters* ViewProviderHole::getEditDialog()
+    return SbVec2f(uCoord, vCoord);
+}
+
+Handle(Geom_Surface) unwrapSurface(const TopoDS_Face& face)
```
[a79c23a](https://github.com/FreeCAD/FreeCAD/pull/22573/commits/a79c23a509c7eb825dd6ca5aedf4968faaa66fcc)

alfrix commented on src/Mod/PartDesign/Gui/TaskHoleParameters.cpp:
```diff
@@ -828,9 +865,13 @@ void TaskHoleParameters::changedObject(const App::Document&, const App::Property
+    if (&Prop == &hole->Threaded || &Prop == &hole->ModelThread || &Prop == &hole->CosmeticThread) {
         updateHoleTypeCombo();
     }
+    else if (&Prop == &hole->ModelThread) {
```
[670a7cb](https://github.com/FreeCAD/FreeCAD/pull/22573/commits/670a7cb6dcc133b8b8ccf88f9d500e1b295232bb)

InstantMuffin commented:
I would love to have this as a checkbox within the dialog creating that thread, with the checkbox output either being memorized for other dialogs, or with a global setting within the options representing a preference.

pierreporte commented:
@InstantMuffin Please open an issue
